File: RUNME.m
%%% project: emgr - EMpirical GRamian Framework ( https://gramian.de )
%%% version: 5.99 (2022-04-13)
%%% authors: Christian Himpe (0000-0003-2194-6754)
%%% license: BSD-2-Clause (opensource.org/licenses/BSD-2-Clause)
%%% summary: RUNME - minimal test script

emgr_VERSION = emgr('version')

% Linear System
A = -eye(4)					% System Matrix
B = [0;1;0;1]					% Input Matrix
C = [0,0,1,1]					% Output Matrix

P = zeros(4,1);				% Parameter Vector
Q = [0.01*ones(4,1),[0.25;0.5;0.75;1.0]];	% Min and Max Parameter Box

f = @(x,u,p,t) A*x + B*u + p;			% (Affine) Linear Vector Field
g = @(x,u,p,t) C*x;				% Linear Output Functional
h = @(x,u,p,t) A'*x + C'*u;			% Adjoint Vector Field

s = [1,4,1];					% System dimension
t = [0.01,1.0];				% Time discretization

% (Empircal) Controllability Gramian
WC = emgr(f,g,s,t,'c',P)

% (Empirical) Observability Gramian
WO = emgr(f,g,s,t,'o',P)

% (Empirical) Cross Gramian
WX = emgr(f,g,s,t,'x',P)

% (Empirical) Linear Cross Gramian
WY = emgr(f,h,s,t,'y',P)

% (Empirical) Sensitivity Gramian
WCWS = emgr(f,g,s,t,'s',Q)

% (Empirical) Identifiability Gramian
WOWI = emgr(f,g,s,t,'i',Q)

% (Empirical) Cross-Identifiability Gramian
WXWJ = emgr(f,g,s,t,'j',Q)

File: py/RUNME.py
#!/usr/bin/env python3

"""
  project: emgr ( https://gramian.de )
  version: 5.99.py (2022-04-13)
  authors: Christian Himpe (0000-0003-2194-6754)
  license: BSD-2-Clause License (opensource.org/licenses/BSD-2-Clause)
  summary: RUNME (Minimal emgr test script)
"""

import numpy as np
from emgr import emgr


print("emgr version: {0}".format(emgr("version")))

# Linear System
A = -np.eye(4)
B = np.array([[0.0,1.0,0.0,1.0]]).T
C = np.array([[0.0,0.0,1.0,1.0]])

P = np.zeros((4,1))
Q = np.array([[0.01,0.25],[0.01,0.5],[0.01,0.75],[0.01,1.0]])

def f(x,u,p,t): return A.dot(x) + B.dot(u) + p
def g(x,u,p,t): return C.dot(x)
def h(x,u,p,t): return A.T.dot(x) + C.T.dot(u)

s = (1,4,1)
t = (0.01,1.0)

# (Empircal) Controllability Gramian
WC = emgr(f,g,s,t,"c",P)
print(WC)

# (Empirical) Observability Gramian
WO = emgr(f,g,s,t,"o",P)
print(WO)

# (Empirical) Cross Gramian
WX = emgr(f,g,s,t,"x",P)
print(WX)

# (Empirical) Linear Cross Gramian
WY = emgr(f,h,s,t,"y",P)
print(WY)

# (Empircal) Controllability Gramian
WC,WS = emgr(f,g,s,t,"s",Q)
print(WC)
print(WS)

# (Empirical) Observability Gramian
WO,WI = emgr(f,g,s,t,"i",Q)
print(WO)
print(WI)

# (Empirical) Cross Gramian
WX,WJ = emgr(f,g,s,t,"j",Q)
print(WX)
print(WJ)

